R/table-object.R

```diff
@@ -950,3 +950,33 @@ print.stobject <- function(x, ...) {
 #' @rdname print.stobject
 #' @export
 names.stobject <- function(x) names(x$data)
+
+
+#' Clone an stobject
+#'
+#' @param x an stobject object.
+#'
+#' @return
+#' A deep copy of `x`.
```

Comment:
> A deep copy of `x`.

I think it'd be good to avoid the word "deep".  Perhaps I'm too influence by Python (which of course has very different semantics), but I'd expect a deep copy to be recursive.  I'm guessing there's no need for it to be recursive in this case, so my suggestion is just to adjust the docs.

<details>
<summary>example</summary>

For a deep copy, I'd expect the inner `aa` environment to also be copied.

``` r
st_clone <- function(x) {
  y <- list2env(as.list(x))
  attributes(y) <- attributes(x)
  y
}

a <- new.env(parent = emptyenv())
a$foo <- 1
aa <- new.env(parent = emptyenv())
aa$bar <- 2
a$aa <- aa

b <- st_clone(a)
b$foo <- b$foo * 10
b$new <- "abc"
b$aa$bar <- b$aa$bar * 10
b$aa$new <- "xyz"


b$foo
#> [1] 10
b$new
#> [1] "abc"
b$aa$bar
#> [1] 20
b$aa$new
#> [1] "xyz"

a$foo
#> [1] 1
a$new
#> NULL
a$aa$bar
#> [1] 20
a$aa$new
#> [1] "xyz"
```

</details>

Reply:
Fair enough. 

- Dropped "deep" from docs
- Using rlang clone function
- Will error if the argument isn't the object we were expecting (that would prevent the `example` you show from happening
- I believe (as the tests and examples show) once we "clone" this, the new object is subject to copy-on-modify behavior; I think some of the underlying objects in the clone and the original still point to the original address, but if you modify it will get copied
``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(pmtables)
library(lobstr)
a <-
  stdata() %>%
  st_new() %>%
  st_center()


b <- st_clone(a)
c <- a

obj_addr(a)
#> [1] "0x143747dd8"
obj_addr(b)
#> [1] "0x1437be480"
obj_addr(c)
#> [1] "0x143747dd8"

obj_addr(a$align)
#> [1] "0x1164ee2e8"
obj_addr(b$align)
#> [1] "0x1164ee2e8"
b$align <- letters

str(a$align)
#> List of 5
#>  $ complete: NULL
#>  $ default : chr "c"
#>  $ update  : list()
#>  $ coltype : chr "p"
#>  $ outer   : chr "none"
#>  - attr(*, "class")= chr "aligncol"
obj_addr(a$align)
#> [1] "0x1164ee2e8"
obj_addr(b$align)
#> [1] "0x14406d430"
```

<sup>Created on 2023-07-14 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>